course7/homeworks/fred/complex-rpn/complex.cc

```diff
@@ -0,0 +1,42 @@
+#include "complex.h"
+
+#include <stdexcept>
+
+Complex Complex::operator+ (const Complex& other) const
+{
+  Complex result;
+  result.real = this->real + other.real;
+  result.imaginary = this->imaginary + other.imaginary;
+  return result;
+}
+
+Complex Complex::operator- (const Complex& other) const
+{
+  Complex result;
+  result.real = this->real - other.real;
+  result.imaginary = this->imaginary - other.imaginary;
+  return result;
+}
+
+Complex Complex::operator* (const Complex& other) const
+{
+  Complex result;
+  result.real = (this->real * other.real) - (this->imaginary * other.imaginary);
+  result.imaginary = (this->imaginary * other.real) + (this->real * other.imaginary);
+  return result;
+}
+
+Complex Complex::operator/ (const Complex& other) const
+{
+  if (other.real == 0.0 && other.imaginary == 0.0) {
+    // TODO check with epsilon?
+    throw std::runtime_error("Division by zero");
```

Comment:
This violates the "pay only fr what you are using" principle. Make this a narrow contract function, and make this user's responsibility to ensure the values are valid.